### PR TITLE
Plugins Management: Use the new endpoint for fetching all site plugins

### PR DESCRIPTION
**Pull request by vitozev**
#### Proposed Changes

In https://github.com/Automattic/wp-calypso/pull/61803, @enejb did a fantastic job using `react-query` for some plugin endpoints, but IMO we should try to split such refactor into much smaller pieces to avoid the possibility of introducing a regression.

Migrating all plugins to react-query would require us to:
- Research if we can replace Redux completely. In a quick chat, @tyxla pointed out that there might be some dynamics in the UI (e.g., clicking "Activate plugin" would be something we currently can't track on the server side, but it's available in the store).
- Define the phases of how this should happen.
- Work on small incremental steps towards the goal. 

What about this PR?

As properly migrating to react-query would require more time, this changeset introduces a new query component, `QueryAllJetpackPlugins` that is used in all places where we need to fetch installed plugins _for all Jetpack sites._ 
The existing `<QueryJetpackPlugins />` component will still be used for fetching plugins for _a single Jetpack site_.

We found some inconsistencies, which were documented in: pbtFFM-29g-p2

**Note:** I'm working on some tests, but I'd like to get this code reviewed while working on them.

#### Testing Instructions
- Checkout PR locally
- Run `yarn start`
- When the build is ready, run `yarn start-jetpack-cloud-p`
- Visit http://calypso.localhost:3000/plugins/manage and verify that the screen loads as expected (hopefully more instantly as well)
- Verify that results match the ones from the production environment. If they don't match, this shouldn't be directly related to this PR, but would be good to investigate the case further.
- Verify that bulk actions (install, activate, enable auto-update, update and remove plugins) still work as expected.
- Visit http://calypso.localhost:3000/plugins/akismet and verify that the screen loads as expected.
- Visit http://calypso.localhost:3000/plugins/browse/popular and verify that the screen loads as expected.
- Visit http://jetpack.cloud.localhost:3001/plugins/manage and verify that the screen loads as expected (and matches what you see on Calypso Blue)
- Visit http://jetpack.cloud.localhost:3001/plugins/akismet and verify that the screen loads as expected.
- Visit http://jetpack.cloud.localhost:3001/plugins/manage/<SITE> and verify that the screen loads as expected.

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Related to 1202518759611394-as-1202534044814282

### CODE REVIEW

**Comment by rcanepa** [comment]
> Verify that results match the ones from the production environment. If they don't match, this shouldn't be directly related to this PR, but would be good to investigate the case further.

I know this is unlikely to be related to these changes, but I'm describing it here so that it can be investigated further.

The number of plugins doesn't match. In concrete, I see two Jetpack plugins that are not visible in production (see the ones highlighted in a red box). I haven't checked the code, but maybe we are hiding/skipping some specific plugins.

image

File: client/state/plugins/installed/actions.js [hunk 1]
```diff
@@ -532,3 +535,32 @@ export function fetchSitePlugins( siteId ) {
...
+		const receivePluginsDispatchSuccess = ( { sites } ) => {
+			dispatch( { type: ALL_PLUGINS_REQUEST_SUCCESS } );
+
+			Object.entries( sites ).map( ( [ siteId, plugins = [] ] ) => {
+				dispatch( receiveSitePlugins( siteId, plugins ) );
+
+				plugins.map( ( plugin ) => {
+					if ( plugin.update && plugin.autoupdate ) {
+						updatePlugin( siteId, plugin )( dispatch );
+					}
```
Comment (rcanepa) [inline client/state/plugins/installed/actions.js:552]: A question out of curiosity. Does this mean that if a plugin needs an update and has auto-update enabled, the frontend will dispatch a request to update the plugin?

File: client/state/plugins/installed/selectors.js [hunk 2]
```diff
@@ -52,6 +52,13 @@ export function isRequestingForSites( state, sites ) {
 	return some( sites, ( siteId ) => isRequesting( state, siteId ) );
 }
 
+export function isRequestingForAllSites( state ) {
+	if ( typeof state.plugins.installed.isRequestingAll === 'undefined' ) {
+		return false;
+	}
+	return state.plugins.installed.isRequestingAll;
```
Comment (rcanepa) [inline client/state/plugins/installed/selectors.js]: Is this possible? I see that this piece of state has a default value of `false`, so I don't think it can be `undefined`.

File: client/state/action-types.ts [hunk 3]
```diff
@@ -619,6 +619,9 @@ export const PLUGINS_RECEIVE = 'PLUGINS_RECEIVE';
 export const PLUGINS_REQUEST = 'PLUGINS_REQUEST';
 export const PLUGINS_REQUEST_FAILURE = 'PLUGINS_REQUEST_FAILURE';
 export const PLUGINS_REQUEST_SUCCESS = 'PLUGINS_REQUEST_SUCCESS';
+export const ALL_PLUGINS_REQUEST = 'ALL_PLUGINS_REQUEST';
```
Comment (tyxla) [inline client/state/action-types.ts]: We usually try to keep this list alphabetized. Also, I'd suggest keeping the `PLUGINS_` prefix here. `PLUGINS_ALL_*` sounds okay.

File: client/state/plugins/installed/README.md [hunk 4]
```diff
@@ -26,6 +26,8 @@ A module for managing installed plugins on connected sites.
 
 ### `isRequestingForSites( state: Object, sites: Array )`
 
+### `isRequestingForAllSites( state: Object )`
```
Comment (tyxla) [inline client/state/plugins/installed/README.md:29]: Let's make sure to document this selector.

File: client/state/plugins/installed/selectors.js (hunk 2 above)
Comment (tyxla) [reply client/state/plugins/installed/selectors.js]: I second this. Should be fine to just return without a check.

File: client/components/data/query-all-jetpack-plugins/index.tsx [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import type { CalypsoDispatch } from 'calypso/state/types';
+import type { AppState } from 'calypso/types';
+
+const request = () => ( dispatch: CalypsoDispatch, getState: AppState ) => {
+	if ( ! isRequestingForAllSites( getState() ) ) {
+		dispatch( fetchAllPlugins() );
+	}
+};
+
+export default function QueryAllJetpackPlugins() {
```
Comment (tyxla) [inline client/components/data/query-all-jetpack-plugins/index.tsx]: The query component's name is a bit ambiguous. Am I requesting all plugins of this site? Or am I requesting the plugins for all sites? Is this about installed plugins or WP.org plugins available to install? IMHO this should be better reflected in the query component name.

File: client/my-sites/plugins/main.jsx [hunk 6]
```diff
@@ -442,7 +442,7 @@ export class PluginsMain extends Component {
 		const content = (
 			<>
 				<DocumentHead title={ pageTitle } />
-				<QueryJetpackPlugins siteIds={ this.props.siteIds } />
+				<QueryAllJetpackPlugins />
```
Comment (tyxla) [inline client/my-sites/plugins/main.jsx]: It feels like overkill to me to request the plugins for all sites (and update plugins on all autoupdate-enabled sites) when accessing the plugins page of a single site (`/plugins/manage/:site`). WDYT?

File: client/my-sites/plugins/plugin-details.jsx [hunk 7]
```diff
@@ -283,7 +281,7 @@ function PluginDetails( props ) {
 		<MainComponent wideLayout>
 			<DocumentHead title={ getPageTitle() } />
 			<PageViewTracker path={ analyticsPath } title="Plugins > Plugin Details" />
-			<QueryJetpackPlugins siteIds={ siteIds } />
+			<QueryAllJetpackPlugins />
```
Comment (tyxla) [inline client/my-sites/plugins/plugin-details.jsx]: It feels like overkill to me to request the plugins for all sites (and update plugins on all autoupdate-enabled sites) when accessing the a single plugin page within the context of a single site (`/plugins/query-monitor/:site`). WDYT?

File: client/my-sites/plugins/main.jsx [hunk 8]
```diff
@@ -569,7 +569,7 @@ export default flow(
 				currentPluginsOnVisibleSites: getPlugins( state, visibleSiteIds, filter ),
 				pluginUpdateCount: pluginsWithUpdates && pluginsWithUpdates.length,
```
Comment (tyxla) [inline client/my-sites/plugins/main.jsx:605]: Should we update all additional areas that fetch or return plugin data for the selected `siteIds` to actually do it for all sites? For example, `pluginsWithUpdates` will still return only the plugins with updates within the `siteIds` prop.

File: client/state/plugins/installed/actions.js [hunk 9]
```diff
@@ -532,3 +535,32 @@ export function fetchSitePlugins( siteId ) {
...
+		};
+
+		const receivePluginsDispatchFail = ( error ) => {
+			dispatch( { type: ALL_PLUGINS_REQUEST_FAILURE, error } );
+		};
+
+		return wpcom.req
+			.get( `/me/sites/plugins` )
+			.then( receivePluginsDispatchSuccess )
+			.catch( receivePluginsDispatchFail );
```
Comment (tyxla) [inline client/state/plugins/installed/actions.js:564]: What kind of error handling are we doing, just swallowing the errors?

File: client/state/plugins/installed/actions.js [hunk 10]
```diff
@@ -532,3 +535,32 @@ export function fetchSitePlugins( siteId ) {
...
+
+export function fetchAllPlugins() {
+	return ( dispatch ) => {
+		dispatch( { type: ALL_PLUGINS_REQUEST } );
+
+		const receivePluginsDispatchSuccess = ( { sites } ) => {
+			dispatch( { type: ALL_PLUGINS_REQUEST_SUCCESS } );
+
+			Object.entries( sites ).map( ( [ siteId, plugins = [] ] ) => {
+				dispatch( receiveSitePlugins( siteId, plugins ) );
```
Comment (tyxla) [inline client/state/plugins/installed/actions.js:547]: Is the site plugin data verified to be in the same format as it was when we were requesting data per site?

File: client/state/plugins/installed/actions.js [hunk 11]
```diff
@@ -532,3 +535,32 @@ export function fetchSitePlugins( siteId ) {
...
 }
+
+export function fetchAllPlugins() {
+	return ( dispatch ) => {
+		dispatch( { type: ALL_PLUGINS_REQUEST } );
+
+		const receivePluginsDispatchSuccess = ( { sites } ) => {
+			dispatch( { type: ALL_PLUGINS_REQUEST_SUCCESS } );
+
+			Object.entries( sites ).map( ( [ siteId, plugins = [] ] ) => {
```
Comment (tyxla) [inline client/state/plugins/installed/actions.js]: Is the default empty array of plugins necessary?

Comment (tyxla) [inline client/state/plugins/installed/actions.js]: Do we need to handle the case when there are no `sites`?

File: client/state/plugins/installed/actions.js [hunk 12]
```diff
@@ -532,3 +535,32 @@ export function fetchSitePlugins( siteId ) {
...
+				dispatch( receiveSitePlugins( siteId, plugins ) );
+
+				plugins.map( ( plugin ) => {
```
Comment (tyxla) [inline client/state/plugins/installed/actions.js]: Seems like we're not really mapping, but rather each-ing, so `forEach` could be the more adequate syntax for this use case.

**Comment by rcanepa** [comment]
After a quick code review I can say that the changes look reasonable and easy to follow. Furthermore, features seem to work as expected as well. However, I'll spend more time tomorrow reviewing everything before giving an official review.

File: client/state/plugins/installed/reducer.js [hunk 13]
```diff
@@ -18,6 +21,27 @@ import { combineReducers, withSchemaValidation } from 'calypso/state/utils';
...
+/**
+ * Returns the updated requesting state after an action has been dispatched.
+ * Requesting state tracks whether a network request is in progress for all
+ * plugins on all sites.
+ *
+ * @param  {object} state  Current state
+ * @param  {object} action Action object
+ * @returns {object}        Updated state
+ */
+export function isRequestingAll( state = false, action ) {
```
Comment (tyxla) [inline client/state/plugins/installed/reducer.js:33]: Do we also need to handle the `isRequesting()` and `isRequestingForSites` selectors?

File: client/state/plugins/installed/selectors.js [hunk 14]
```diff
@@ -60,6 +67,12 @@ export function getPlugins( state, siteIds, pluginFilter ) {
 				return memo;
 			}
 
+			// We currently support fetching plugins per site and also fetching all plugins
+			// in bulk, aiming to optimize the UX in some flows.
+			if ( isRequestingForAllSites( state ) ) {
```
Comment (tyxla) [inline client/state/plugins/installed/selectors.js:69]: Should this optimization happen in some other places too? Like when we're grabbing a single site plugin for example?

File: client/state/plugins/installed/actions.js [hunk 15]
```diff
@@ -532,3 +535,32 @@ export function fetchSitePlugins( siteId ) {
...
+				} );
+			} );
+		};
+
+		const receivePluginsDispatchFail = ( error ) => {
+			dispatch( { type: ALL_PLUGINS_REQUEST_FAILURE, error } );
+		};
+
+		return wpcom.req
+			.get( `/me/sites/plugins` )
```
Comment (tyxla) [inline client/state/plugins/installed/actions.js:562]: By any chance is this endpoint using cached data? I went to `/plugins/query-monitor` and installed the plugin on site that didn't previously have it installed, and when I briefly refreshed after the installation, the plugin didn't appear to be installed. It appeared on the next refresh though. Seems like a regression if we're not updating the data immediately. Any idea what might have happened there?

**Review by tyxla — COMMENTED** [review]
Thanks for the iterative approach. The plugins area is complex and there are so many scenarios and test cases to consider when making changes, so we should be careful.

I've left some thoughts and feedback, hope that helps.

In any case, let's make sure to thoroughly test all different scenarios and combinations of:

* Single site / all sites
* Single plugin / all plugins
* WP.org plugins

There might be some quirky behaviors that we want to ensure don't happen, for example: what happens if we're installing a plugin in the background, and then we navigate to `/plugins/manage/:site`  and a refetch happens? Will it override the data because it will not be aware about the plugin we just installed?

File: client/components/data/query-all-jetpack-plugins/index.tsx (hunk 5 above)
Comment (vitozev) [reply client/components/data/query-all-jetpack-plugins/index.tsx]: Thanks! What do you think about `QueryAllMyJetpackSitesPlugins`?

File: client/state/plugins/installed/README.md (hunk 4 above)
Comment (vitozev) [reply client/state/plugins/installed/README.md:29]: Thanks! The selector was documented in 839e7834a1f19087233fee5137dc89ec9bb1a5e2

File: client/state/action-types.ts (hunk 3 above)
Comment (vitozev) [reply client/state/action-types.ts]: Thanks! I agree `PLUGINS_ALL_*` sounds better and I've updated this in 839e7834a1f19087233fee5137dc89ec9bb1a5e2

File: client/components/data/query-all-jetpack-plugins/index.tsx (hunk 5 above)
Comment (tyxla) [reply client/components/data/query-all-jetpack-plugins/index.tsx]: Sounds good to me! We could ditch `My` though.

Comment (vitozev) [reply client/components/data/query-all-jetpack-plugins/index.tsx]: I agree! Renamed in 8607109ef88edccbc285de74eee3b39809403ea9

File: client/state/plugins/installed/selectors.js (hunk 2 above)
Comment (vitozev) [reply client/state/plugins/installed/selectors.js]: Thanks! Fixed in f47988bd6bbdc03b69064032f23a2db030cb49ff

File: client/state/plugins/installed/actions.js (hunk 15 above)
Comment (vitozev) [reply client/state/plugins/installed/actions.js:562]: We are using `get_option` (AFAIK, it caches the result, so any further call for that option on the same HTTP request won't hit the DB), but there's no caching layer implemented.

I've installed a random plugin from `http://calypso.localhost:3000/plugins/:site` and waited for the activation request to finish. Then I refreshed the `http://calypso.localhost:3000/plugins/manage/:site` (in a new tab), and my plugin appeared in the list.

Am I testing this correctly?

In theory, there could be a slight delay because we need that sync action (after the plugin is installed) to be triggered, but I'd say this shouldn't be noticed in most cases.

File: client/my-sites/plugins/plugin-details.jsx (hunk 7 above)
Comment (vitozev) [reply client/my-sites/plugins/plugin-details.jsx]: Oh, you're absolutely right! Thanks.

Fixed in fba75ea7f0cf023bc061edcf7975cc5c65142abb

File: client/my-sites/plugins/main.jsx (hunk 6 above)
Comment (vitozev) [reply client/my-sites/plugins/main.jsx]: Thanks! Fixed in fba75ea7f0cf023bc061edcf7975cc5c65142abb

File: client/state/plugins/installed/actions.js (hunk 11 above)
Comment (vitozev) [reply client/state/plugins/installed/actions.js:546]: The API will always return an empty array, even if there are no sites. Maybe we can leave it as-is in this case?

File: client/state/plugins/installed/actions.js (hunk 9 above)
Comment (vitozev) [reply client/state/plugins/installed/actions.js:564]: Yes, it looks like the `PLUGINS_REQUEST_FAILURE` is passing the actual message through, so I did the same for the new endpoint, but I created a new card in Asana to fix this.

1202518759611394-as-1202789398406899

File: client/state/plugins/installed/actions.js (hunk 12 above)
Comment (vitozev) [reply client/state/plugins/installed/actions.js]: Good catch! Fixed in 9936967d79fd3fe3c8625446d5ab67d167317e74

File: client/state/plugins/installed/actions.js (hunk 10 above)
Comment (vitozev) [reply client/state/plugins/installed/actions.js:547]: Yes, the data returned by the new endpoint has the same format as the per-site endpoint.

File: client/state/plugins/installed/actions.js (hunk 11 above)
Comment (vitozev) [reply client/state/plugins/installed/actions.js]: The API would return an empty array even if there are no sites.

Comment (vitozev) [reply client/state/plugins/installed/actions.js]: Thanks! It's not needed, actually. Fixed in 3417d8de4170096f1fcad1611afc74daffa0d367

File: client/my-sites/plugins/main.jsx (hunk 8 above)
Comment (vitozev) [reply client/my-sites/plugins/main.jsx:605]: In general, the `siteIds` in the UI and the sites returned from the backend should match, but I agree it would be better to update all additional areas as well to avoid any inconsistency.

What do you think about adding a new `getAllPlugins()` (or similar) selector that will be used when there's no selected site?

File: client/state/plugins/installed/actions.js (hunk 15 above)
Comment (tyxla) [reply client/state/plugins/installed/actions.js:562]: Sounds like you're testing it correctly 👍 

However, `get_option()` runs on a shadow (AKA cache) site, and the data there is populated by sync. There's no guarantee that once you perform a plugin action, it will be synced immediately to the cache site. The delay can be substantial when there's a lot of action going on on the site and there's a large queue of actions to be synced. It may be running quickly on your test site, but that's likely because there are no queued actions for syncing, and the plugin actions get synced to the shadow site immediately.

TBH this seems problematic to me, as it can lead to inconsistent behavior and bad UX.

File: client/my-sites/plugins/main.jsx (hunk 8 above)
Comment (tyxla) [reply client/my-sites/plugins/main.jsx:605]: Such a selector could solve the issue, but to me the root of the problem is that there are a bunch of components and subcomponents in the tree that potentially consume different sets of `siteId` or `siteIds` and that could lead to weird UX issues and errors.

Comment (vitozev) [reply client/my-sites/plugins/main.jsx:605]: Initially, I thought we could replace all of them, but that would be a lot of work, and I'm not sure if we should include this in the scope of this PR.

---

Currently, we fetch the sites from the `me/sites` endpoint, then we have the [getSelectedOrAllSites](https://github.com/Automattic/wp-calypso/blob/7655b6d1f326371990851da10ebb97f133fbdc17/client/state/selectors/get-selected-or-all-sites-with-plugins.js#L16) selector that does some additional filtering.

The returned IDs from the above should match the ones returned _by the new endpoint_: active Jetpack sites where the current user is capable of managing plugins.

Can you think of any reason why the returned sites would differ in both situations?

Comment (tyxla) [reply client/my-sites/plugins/main.jsx:605]: One example I can think of is this page: https://wordpress.com/plugins/manage

There you can still select one or more plugins to perform bulk actions on, and that would work only with the sites where that plugin is installed on. In that scenario, the list of sites can, and most likely will, differ from the list of all Jetpack sites.

File: client/state/plugins/installed/actions.js (hunk 15 above)
Comment (vitozev) [reply client/state/plugins/installed/actions.js:562]: Thanks for all your feedback! To avoid going back and forth here, we moved the discussion [here](pbtFFM-23q-p2#comment-2908).

As the sync times are much better now, we can leave this as-is and monitor it closely to see if there will be any inconsistencies.

File: client/my-sites/plugins/main.jsx (hunk 8 above)
Comment (tyxla) [reply client/my-sites/plugins/main.jsx:605]: Actually, disregard this message. I realized that the state management for managing selected plugins for bulk editing is not handled in the main component, but rather by the `PluginsList`, a level down, so we should be good here. Seems like the above shouldn't be a concern 👍

File: client/state/plugins/installed/actions.js (hunk 1 above)
Comment (vitozev) [reply client/state/plugins/installed/actions.js:552]: That's right. I'm not sure about why it works this way, though. @tyxla, do you have any thoughts to share here? I'm curious what would happen if the user never visits the plugin management UI, do we have any fallback mechanism for running auto-updates?

Comment (tyxla) [reply client/state/plugins/installed/actions.js:552]: We could remove that if we want to. We're currently triggering an async job to auto-update plugins after each sync action, look for `maybe_autoupdate` in the WP.com codebase and you'll find it.

I believe this is here because the async job is not guaranteed to run anytime soon, though. Whether this way to force autoupdate is a good UX, that's arguable.

File: client/state/plugins/installed/reducer.js (hunk 13 above)
Comment (vitozev) [reply client/state/plugins/installed/reducer.js:33]: The `isRequesting()` is handled [here](https://github.com/Automattic/wp-calypso/blob/8458f5119c43cfb194f994352d71654851d4aced/client/state/plugins/installed/reducer.js#L24). It looks like `isRequestingForSites()` simply uses it, so I guess we should be good to go. FWIW, these selectors weren't introduced in this PR.

**Review by rcanepa — APPROVED** [review]
I focused most on ensuring the changes work as expected since Marin focused on the code. The result of my review is that things work as described:

- I don't see any difference between development and production on Calypso Blue (I can't compare Calypso Green and Calypso Blue since the former now has a different UI for plugin management). Specifically, the general and details view look as expected.
- Bulk actions work as usual.
- I didn't see any regression.

In summary, considering all pieces of feedback have been addressed, and the changes work as expected, I approve these changes.

File: client/state/plugins/installed/selectors.js (hunk 14 above)
Comment (vitozev) [reply client/state/plugins/installed/selectors.js:69]: It looks like `getPluginOnSite` (for example) reuses `getPlugins` behind the scenes, so I think we should be good here. 

I also tested various flows, and I couldn't find any regression in terms of missing loading indicators (or something similar) because of that change.

**Comment by vitozev** [comment]
@rcanepa, thank you for your review! I added one card to investigate why some test plugins are returned from the new endpoint.

In short, if you open your mobile app, you most likely will see the same plugin list as the one returned by our new endpoint.

Card: 1202518759611394-as-1202882177960021

**Comment by vitozev** [comment]
@tyxla, thank you for all your great feedback here! I think I've addressed everything. Let me know if you have the bandwidth to do a quick review of everything that was changed before we proceed with the merging of this PR.

FWIW, we will write some new unit tests in a [separate PR](1202518759611394-as-1202882177960012) in the following days.

**Comment by tyxla** [comment]
> @tyxla, thank you for all your great feedback here! I think I've addressed everything. Let me know if you have the bandwidth to do a quick review of everything that was changed before we proceed with the merging of this PR.
> 
> FWIW, we will write some new unit tests in a [separate PR](1202518759611394-as-1202882177960012) in the following days.

I can take another look at it tomorrow if that is not too late for you folks.

File: client/my-sites/plugins/main.jsx [hunk 16]
```diff
@@ -460,7 +465,11 @@ export class PluginsMain extends Component {
 		const content = (
 			<>
 				<DocumentHead title={ pageTitle } />
-				<QueryJetpackPlugins siteIds={ this.props.siteIds } />
+				{ selectedSite ? (
+					<QueryJetpackPlugins siteIds={ [ selectedSite.ID ] } />
+				) : (
+					<QueryAllJetpackSitesPlugins />
+				) }
```
Comment (tyxla) [inline client/my-sites/plugins/main.jsx:472]: Given how many times we're repeating this logic, is it worth extracting it to another component where we could reuse it?

File: client/state/action-types.ts [hunk 17]
```diff
@@ -624,6 +624,9 @@ export const PLUGINS_RECEIVE = 'PLUGINS_RECEIVE';
 export const PLUGINS_REQUEST = 'PLUGINS_REQUEST';
 export const PLUGINS_REQUEST_FAILURE = 'PLUGINS_REQUEST_FAILURE';
 export const PLUGINS_REQUEST_SUCCESS = 'PLUGINS_REQUEST_SUCCESS';
+export const PLUGINS_ALL_REQUEST = 'PLUGINS_ALL_REQUEST';
```
Comment (tyxla) [inline client/state/action-types.ts]: Do you mind sorting these so all action types continue to be in alphabetical order?

**Review by tyxla — APPROVED** [review]
Left a few minor suggestions, but nothing blocking, code looks great 👍 Thanks 🚀

File: client/my-sites/plugins/main.jsx (hunk 16 above)
Comment (vitozev) [reply client/my-sites/plugins/main.jsx:472]: Thanks! I added a new card and we will address this in a separate PR: 1202518759611394-as-1202889173272225

File: client/state/action-types.ts (hunk 17 above)
Comment (vitozev) [reply client/state/action-types.ts]: Thanks! Fixed in 507aa423b5f3e310d59aee2b4c8494ac112cd791
